Will now persist item mapping for external item

nnguyen259 opened the pull request:
Bugfix branch for chapter 2 release

Bugfixes include:
- External Item item mapping is now persisted
- Fix bug where player id was used for rule quantity in locations.py

nnguyen259 committed:
Will now persist item mapping for external item

tdkollins commented on worlds/tits_the_3rd/client/client.py:
```diff
@@ -374,6 +377,19 @@ def install_game_mod(self):
                 if player_id_file.read() == f"{self.auth}-{self.seed_name}":
                     logger.info("Player has not changed. Skip installing patch")
+                    try:
+                        with open(game_dir / "data/ED6_DT22/item_id_mapping.json") as mapping_file:
```
Nit, the rest of this file uses os.path.join

nnguyen259 committed:
Fix client issue when crafts are not shuffled and remove some debug log

nnguyen259 committed:
os.path.join instead of Path concatation

nnguyen259 committed:
Fix the issue again?

Commit:
Fix bug where gen uses player id for item requirement quantity

nnguyen259 committed:
Add mapping for crafts that did not get shuffled

Commit:
Fix bug where upgraded craft IDs did not get passed to patch data

nnguyen259 committed:
Fix day grancel port chest and client issue with alias

nnguyen259 committed:
Fix Unicode Error and log everything to the info stream instead of error

nnguyen259 committed:
Fix some item name and id

nnguyen259 committed:
Fix several gen and unittest issues

tdkollins commented on worlds/tits_the_3rd/__init__.py:
```diff
@@ -124,14 +124,20 @@ def setup_characters(self, itempool: Counter[str]) -> Counter[str]:
             for character in character_list:
                 if character in default_character_to_location:
                     location_name = default_character_to_location[character]
+                    if self.options.name_spoiler_option:
+                        real_location_name = scrub_spoiler_data(location_name)
+                    else:
+                        real_location_name = location_name
```
It seems like we do this block a lot. Lets make a wrapper for this? So we can just do:
`self.multiworld.get_location(resolve_name(location_name), ...)`

nnguyen259 committed:
Add and/or fix the following:

Fix 2 more items than locations
Fix some more item names
Add victory rule to logic
Add victory condition option

tdkollins commented on worlds/tits_the_3rd/__init__.py:
```diff
@@ -223,6 +226,13 @@ def pre_fill(self) -> None:
+        # For now hard code beating Bennu as victory
+        chapter_1_item = self.create_item(ItemName.chapter_1_cleared)
+        chapter_2_item = self.create_item(ItemName.chapter_2_cleared)
+        # TODO: change victory condition
```
Do we still need this comment?

tdkollins commented on worlds/tits_the_3rd/__init__.py:
```diff
@@ -223,6 +226,13 @@ def pre_fill(self) -> None:
         """Define items for Trails in the Sky the 3rd AP"""
+        # For now hard code beating Bennu as victory
+        chapter_1_item = self.create_item(ItemName.chapter_1_cleared)
+        chapter_2_item = self.create_item(ItemName.chapter_2_cleared)
```
I'm thinking this is fine as is if we want to be lazy... but just to bring this up, I think the more archipelago-esque way of achieving victory conditions would be through [events](https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/world%20api.md#events-or-generation-only-itemslocations).

The world API docs show [an example](https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/world%20api.md#events-or-generation-only-itemslocations) of placing a locked victory event in the set_rules method. Maybe we should follow this pattern?

tdkollins replied:
I also just re-read this in the docs... lol 😅 


> pre_fill(self), fill_hook(self) and post_fill(self) called to modify item placement before, during, and after the regular fill process; all finishing before generate_output. Any items that need to be placed during pre_fill should not exist in the itempool, and if there are any items that need to be filled this way, but need to be in state while you fill other items, they can be returned from get_pre_fill_items.
